packages/ui-app/src/InputNumber.tsx

```diff
@@ -0,0 +1,159 @@
+};
+
+class InputNumber extends React.PureComponent<Props, State> {
+  state: State;
+
+  constructor (props: Props) {
+    super(props);
+
+    this.state = {
+      error: '',
+      warn: '',
+      info: '',
+      isPreKeyDown: false
+    };
+  }
+
+  render () {
+    const { className, defaultValue, isError, label, maxLength, placeholder, style, withLabel } = this.props;
+    const { error, info, warn } = this.state;
+
+    return (
+      <div
+        className={classes('ui--InputNumber', className)}
+        style={style}
+      >
+        <Input
+          className={className}
+          defaultValue={defaultValue || '0'}
+          error={error}
+          info={info}
+          isError={isError}
+          label={label}
+          maxLength={maxLength}
+          onChange={this.onChange}
+          onKeyDown={this.onKeyDown}
+          onKeyUp={this.onKeyUp}
+          placeholder={placeholder}
+          style={style}
+          type='text'
+          warn={warn}
+          withLabel={withLabel}
+        />
+      </div>
+    );
+  }
+
+  onChange = (value: string): void => {
+    const { onChange, t } = this.props;
+
+    try {
+      const { isValid, errorMessage, infoMessage, warnMessage } = isValidBalance(value, t);
+
+      this.setState({
+        error: !isValid && errorMessage ? errorMessage : '',
+        info: isValid && infoMessage ? infoMessage : '',
+        warn: isValid && warnMessage ? warnMessage : ''
+      });
+
+      let valueBN: BN;
+      valueBN = new BN(value || '0');
+
+      if (!onChange) {
+        return;
+      }
+
+      onChange({
+        isValid,
+        value: valueBN
+      });
+    } catch (error) {
+      console.error(error);
+    }
+  }
+
+  onKeyDown = (event: React.KeyboardEvent<Element>): void => {
+    const { t } = this.props;
+    const { isPreKeyDown } = this.state;
+
+    // only allow user balance input to contain one instance of '.' for decimals.
+    // prevent use of shift key
+    if (
+      (keydown.isDuplicateDecimalPoint(event.key, (event.target as HTMLInputElement).value)) ||
+      (keydown.isShift(event.shiftKey))
+    ) {
+      event.preventDefault();
+      return;
+    }
+
+    if (KEYS_ALLOWED.includes(event.key)) {
+      return;
+    }
+
+    if (KEYS_PRE.includes(event.key)) {
+      this.setState({ isPreKeyDown: true });
+    }
+
+    // allow users to to use cut/copy/paste combinations, but not non-numeric letters individually
+    if (
+      (keydown.isSelectAll(event.key, isPreKeyDown)) ||
+      (keydown.isCut(event.key, isPreKeyDown)) ||
+      (keydown.isCopy(event.key, isPreKeyDown)) ||
+      (keydown.isPaste(event.key, isPreKeyDown))
+    ) {
+      return;
+    }
+
+    // prevent input of non-integer values (allow numeric including from keyboards with numpad)
+    if (keydown.isNonNumeric(event.key)) {
+      this.setState({
+        error: t('balance.error.format', {
+          defaultValue: 'Balance in DOTs must be a positive number'
```

Comment:
Can't we just reject the keys outright? As a normal `type='number'` HTML input behaves. (Not a major issue, but not allowing is always better than allowing and then saying "you are doing it wrong")

Reply:
EDIT: This will make the `isValid` pass-up not needed as well since we will always have a valid value here.

Reply:
Additionally, This is a number Input, not a Balance input - so a number can be anything, not just a DOT. (Also remove the DOT references everywhere, for substrate chains they have their own tokens)